src/webgpu/capability_info.ts

```diff
@@ -494,65 +478,134 @@ const kBindingKind: {
+  } else {
+    return { usage: GPUConst.TextureUsage.SAMPLED, ...kBindingKind.sampledTexMS, ...kValidStagesAll, };
+  }
+}
+export const kTextureSampleTypes = [
+  'float',
+  'unfilterable-float',
+  'depth',
+  'sint',
+  'uint',
+] as const;
+assertTypeTrue<TypeEqual<GPUTextureSampleType, typeof kTextureSampleTypes[number]>>();
+
+export function storageTextureBindingTypeInfo(d: GPUStorageTextureBindingLayout) {
+  /* prettier-ignore */
+  switch (d.access) {
+    case 'read-only':  return { usage: GPUConst.TextureUsage.STORAGE, ...kBindingKind.storageTex, ...kValidStagesAll,          };
+    case 'write-only': return { usage: GPUConst.TextureUsage.STORAGE, ...kBindingKind.storageTex, ...kValidStagesStorageWrite, };
+  }
+}
+export const kStorageTextureAccessValues = ['read-only', 'write-only'] as const;
+assertTypeTrue<TypeEqual<GPUStorageTextureAccess, typeof kStorageTextureAccessValues[number]>>();
+
+export type BGLEntry = Omit<GPUBindGroupLayoutEntry, 'binding' | 'visibility'>;
+export function texBindingTypeInfo(e: BGLEntry) {
+  if (e.texture !== undefined) return textureBindingTypeInfo(e.texture);
+  if (e.storageTexture !== undefined) return storageTextureBindingTypeInfo(e.storageTexture);
+  unreachable();
+}
+export function bindingTypeInfo(e: BGLEntry) {
+  if (e.buffer !== undefined) return bufferBindingTypeInfo(e.buffer);
+  if (e.texture !== undefined) return textureBindingTypeInfo(e.texture);
+  if (e.sampler !== undefined) return samplerBindingTypeInfo(e.sampler);
+  if (e.storageTexture !== undefined) return storageTextureBindingTypeInfo(e.storageTexture);
+  unreachable('GPUBindGroupLayoutEntry has no BindingLayout');
+}
 
-// All binding types (merged from above)
+export function bufferBindingEntries(includeUndefined: boolean): readonly BGLEntry[] {
+  return [
+    ...(includeUndefined ? [{ buffer: { type: undefined } }] : []),
+    { buffer: { type: 'uniform' } },
+    { buffer: { type: 'storage' } },
+    { buffer: { type: 'read-only-storage' } },
+  ] as const;
+}
+export function samplerBindingEntries(includeUndefined: boolean): readonly BGLEntry[] {
+  return [
+    ...(includeUndefined ? [{ sampler: { type: undefined } }] : []),
+    { sampler: { type: 'comparison' } },
+    { sampler: { type: 'filtering' } },
+    { sampler: { type: 'non-filtering' } },
+  ] as const;
+}
+export function textureBindingEntries(includeUndefined: boolean): readonly BGLEntry[] {
+  return [
+    ...(includeUndefined ? [{ texture: { multisampled: undefined } }] : []),
+    { texture: { multisampled: false } },
+    { texture: { multisampled: true } },
+  ] as const;
+}
+export function storageTextureBindingEntries(format: GPUTextureFormat): readonly BGLEntry[] {
+  return [
+    { storageTexture: { access: 'read-only', format } },
+    { storageTexture: { access: 'write-only', format } },
+  ] as const;
+}
+export function sampledAndStorageBindingEntries(
```

Comment:
The naming stuff for textureBindingEntries, storageTextureBindingEntries, and sampledAndStorageBindingEntries sounds a little bit weird for me. I may like to use sampledTextureBindingEntries, storageTextureBindingEntries, and TextureBindingEntries (including sampledTexture and storageTexture). But the spec does use TextureBindingLayout and StorageTextureBindingLayout, which is a little weird for me too (Why not use SampledTextureBindingLayout and StorageTextureBindingLayout?). 

I assume that You and other folks in WebGPU WG have already discussed this and chose the right names, though.

Reply:
Makes sense to make this more explicit in this PR, will do that. As for the spec, sampled textures are by far the most common case, so I think we just wanted to make "texture" look like the more common case over "storageTexture". There's certainly an argument that both should be explicit, but I don't think it's worth another change at this point...

Reply:
I revisited the discussions when (Google's) Brandon submitted a PR to change BGL. There were a lot of discussions. A few guys did raise the concern about these names like GPUTextureBindingLayout and GPUStorageTextureBindingLayout at that time. But the WG chose the current names. It's fine. I will not waste my time to submit a PR to change the names.